// mountable-hypertrie/lib/messages.js
// This file is auto generated by the protocol-buffers compiler

/* eslint-disable quotes */
/* eslint-disable indent */
/* eslint-disable no-redeclare */
/* eslint-disable camelcase */

// Remember to `npm install --save protocol-buffers-encodings`
var encodings = require('protocol-buffers-encodings')
var varint = encodings.varint
var skip = encodings.skip

var Mount = exports.Mount = {
  buffer: true,
  encodingLength: null,
  encode: null,
  decode: null
}

defineMount()

function defineMount () {
  var enc = [
    encodings.bytes,
    encodings.string,
    encodings.varint
  ]

  Mount.encodingLength = encodingLength
  Mount.encode = encode
  Mount.decode = decode

  function encodingLength (obj) {
    var length = 0
    if (!defined(obj.key)) throw new Error("key is required")
    var len = enc[0].encodingLength(obj.key)
    length += 1 + len
    if (!defined(obj.localPath)) throw new Error("localPath is required")
    var len = enc[1].encodingLength(obj.localPath)
    length += 1 + len
    if (defined(obj.version)) {
      var len = enc[2].encodingLength(obj.version)
      length += 1 + len
    }
    if (defined(obj.remotePath)) {
      var len = enc[1].encodingLength(obj.remotePath)
      length += 1 + len
    }
    if (defined(obj.hash)) {
      var len = enc[0].encodingLength(obj.hash)
      length += 1 + len
    }
    return length
  }

  function encode (obj, buf, offset) {
    if (!offset) offset = 0
    if (!buf) buf = Buffer.allocUnsafe(encodingLength(obj))
    var oldOffset = offset
    if (!defined(obj.key)) throw new Error("key is required")
    buf[offset++] = 10
    enc[0].encode(obj.key, buf, offset)
    offset += enc[0].encode.bytes
    if (!defined(obj.localPath)) throw new Error("localPath is required")
    buf[offset++] = 18
    enc[1].encode(obj.localPath, buf, offset)
    offset += enc[1].encode.bytes
    if (defined(obj.version)) {
      buf[offset++] = 24
      enc[2].encode(obj.version, buf, offset)
      offset += enc[2].encode.bytes
    }
    if (defined(obj.remotePath)) {
      buf[offset++] = 34
      enc[1].encode(obj.remotePath, buf, offset)
      offset += enc[1].encode.bytes
    }
    if (defined(obj.hash)) {
      buf[offset++] = 42
      enc[0].encode(obj.hash, buf, offset)
      offset += enc[0].encode.bytes
    }
    encode.bytes = offset - oldOffset
    return buf
  }

  function decode (buf, offset, end) {
    if (!offset) offset = 0
    if (!end) end = buf.length
    if (!(end <= buf.length && offset <= buf.length)) throw new Error("Decoded message is not valid")
    var oldOffset = offset
    var obj = {
      key: null,
      localPath: "",
      version: 0,
      remotePath: "",
      hash: null
    }
    var found0 = false
    var found1 = false
    while (true) {
      if (end <= offset) {
        if (!found0 || !found1) throw new Error("Decoded message is not valid")
        decode.bytes = offset - oldOffset
        return obj
      }
      var prefix = varint.decode(buf, offset)
      offset += varint.decode.bytes
      var tag = prefix >> 3
      switch (tag) {
        case 1:
        obj.key = enc[0].decode(buf, offset)
        offset += enc[0].decode.bytes
        found0 = true
        break
        case 2:
        obj.localPath = enc[1].decode(buf, offset)
        offset += enc[1].decode.bytes
        found1 = true
        break
        case 3:
        obj.version = enc[2].decode(buf, offset)
        offset += enc[2].decode.bytes
        break
        case 4:
        obj.remotePath = enc[1].decode(buf, offset)
        offset += enc[1].decode.bytes
        break
        case 5:
        obj.hash = enc[0].decode(buf, offset)
        offset += enc[0].decode.bytes
        break
        default:
        offset = skip(prefix & 7, buf, offset)
      }
    }
  }
}

function defined (val) {
  return val !== null && val !== undefined && (typeof val !== 'number' || !isNaN(val))
}
